File: mini-project/product-order-service/src/main/java/com/example/productorderservice/product/application/service/ProductService.java
package com.example.productorderservice.product.application.service;

import com.example.productorderservice.product.application.port.ProductPort;
import com.example.productorderservice.product.domain.Product;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ProductService {
    private final ProductPort productPort;

    public ProductService(ProductPort productPort) {
        this.productPort = productPort;
    }


    @Transactional
    public void addProduct(AddProductRequest request) {
        final Product product = new Product(request.name(), request.price(), request.discountPolicy());

        productPort.save(product);
    }

    public GetProductResponse getProduct(Long productId) {
        final Product product = productPort.getProduct(productId);

        return new GetProductResponse(
                product.getId(),
                product.getName(),
                product.getPrice(),
                product.getDiscountPolicy()
        );
    }

    @Transactional
    public void updateProduct(Long productId, UpdateProductRequest request) {
        final Product product = productPort.getProduct(productId);
        product.update(request.name(), request.price(), request.discountPolicy());
    }
}

File: distributed-transaction/monolithic/src/main/kotlin/org/example/monolithic/product/application/ProductService.kt
package org.example.monolithic.product.application

import org.example.monolithic.product.infrastructure.ProductRepository
import org.springframework.stereotype.Service

@Service
class ProductService(
    private val productRepository: ProductRepository,
) {

    fun buy(productId: Long, quantity: Long): Long {
        val product = productRepository.findById(productId).orElseThrow { throw RuntimeException("데이터가 존재하지 않습니다.") }

        val totalPrice = product.calculatePrice(quantity)
        product.buy(quantity)

        productRepository.save(product)

        return totalPrice
    }
}